tests/core/async_module/test_demo_middleware_conversion.py

```diff
@@ -0,0 +1,100 @@
+import pytest
+
+from eth_utils import (
+    is_checksum_address,
+)
+from hexbytes import (
+    HexBytes,
+)
+
+from web3 import (
+    EthereumTesterProvider,
+    Web3,
+)
+from web3._utils.method_callers import (
+    MethodBinder,
+    PropertyBinder,
+    async_caller,
+)
+from web3._utils.method_formatters import (
+    lookup_pythonic_output_formatter,
+    pythonic_input_formatter,
+)
+from web3._utils.toolz import (
+    assoc,
+    compose,
+    identity,
+)
+from web3._utils.transactions import (
+    get_buffered_gas_estimate,
+)
+from web3.module import (
+    Module,
+)
+
+MIDDLEWARES = []
+
+
+def send_transaction_input_formatter(module, transaction):
+    if 'from' not in transaction and is_checksum_address(module.defaultAccount):
+        transaction = assoc(transaction, 'from', module.defaultAccount)
+
+    # TODO: move gas estimation in middleware
+    if 'gas' not in transaction:
+        transaction = assoc(
+            transaction,
+            'gas',
+            get_buffered_gas_estimate(module.web3, transaction),
+        )
+    return 'eth_sendTransaction', [transaction]
+
+
+class Eth(Module):
+    defaultAccount = ''
+    accounts = PropertyBinder(lambda _: ("eth_accounts", []), identity)
+    sendTransaction = MethodBinder(
+        compose(pythonic_input_formatter, send_transaction_input_formatter),
+        lookup_pythonic_output_formatter('eth_sendTransaction'))
```

Comment:
There are 3 things happening here and maybe this would be cleaner if they were separate arguments.

1. formatting the module method parameters and selecting the rpc method
2. formatting the rpc method params
3. formatting the results.

Would a middleware-esq structure for 2 & 3 be better here?  Or is it nice to have all of the formatting visible up front in the method.  Im slightly in favor of having the formatters up front like this.  It could be made even more explicit without the lookup functions:

```python
sendTransaction = MethodBinder(
   method_formatter=send_transaction_formatter,
   parameter_formatter=identity,
   results_formatter=to_hexbytes(32))

Reply:
First impression: when selecting the method, it may be more natural to depend on the pre-formatted parameters, or the post-formatted parameters. This approach locks is into one or the other. So I'm slightly against, but am very convincable, if I see some examples that make code reuse cleaner.